### ISSUE TITLE
Respect format in canonical links

### ISSUE THREAD

**Issue by matthiasrohmer** [ampproject/amp.dev#2588]
via @tharders: our `<link rel="canonical">` elements currently don't share a format GET parameter and therefore always fallback to *websites*.

This is wrong for documents that shouldn't actually be available for *websites*.

**Comment by sebastianbenz** [comment]
This is tricky. Not sure what the correct approach is. Format specific URL if there's only one format defined? Otherwise pick a random one?

**Comment by matthiasrohmer** [comment]
Indeed. The idea was to use the first one the document is available for though this might not be the most accurate choice.

https://github.com/ampproject/amp.dev/blob/30346f33701ff2aac37c97d1421ba94543af1329/frontend/templates/layouts/default.j2#L47

would then change to something like `<link rel="canonical" href="{{ doc.url }}?format={{ (doc.formats or [])[0] }}"/>`.

**Comment by sebastianbenz** [comment]
I'd suggest:

* use no format specifier if the doc is relevant to all formats. Page will be rendered for `websites`, but we have the option to change the default whenever we want.
* use `formats[0]` if `formats.length === 1`
* use format `websites` if `formats.length > 1` and `formats.includes('websites')` 
* use format `formats[0]` if `formats.length > 1` and `!formats.includes('websites')` 


Sidenote: there's so much of our domain knowledge encoded in template literals. I'd be nice to have these in a separate helper file to ease maintenance and reuse.

**Comment by tharders** [comment]
I would suggest to use the first supported format as default format if the document defined a list of supported formats.
It would then no longer be possible to get the page in an "invalid" format by removing the parameter.

**Comment by tharders** [comment]
The more specific rules that @sebastianbenz suggested are better than simply using the first defined format like I said. But we should use them to select the default format and not to select the format for the canonical link.

**Comment by sebastianbenz** [comment]
Good point: @tharders is right the default canonical link should be clean.

**Comment by fstanis** [comment]
In addition to fixing the canonical, I think we should also add a redirect to `?format=websites` for links that don't have a format.

https://amp.dev/documentation/guides-and-tutorials/start/create/ and https://amp.dev/documentation/guides-and-tutorials/start/create/?format=websites being identical is not a good idea.

**Comment by tharders** [comment]
@fstanis The default format would only be websites if the page supports the websites format.
In that case would you still see a problem that the link without a format parameter does not redirect?

**Comment by fstanis** [comment]
The way I see it (correct me if I'm wrong):

- "No format specified" always means websites.
- This fix only refers to the canonical page `<link>`. In other words, the user will still be able to land on e.g. an ads-only page, but the canonical will then be `?format=ads`.

I have two problems with this:

- `?format=x` "sticks", so subsequent navigation may take the user to the wrong place if no format is specified on the initial page they land on.
- Having two URLs with the exact same content is [duplicate content](https://support.google.com/webmasters/answer/66359) and I see no reason not to have a redirect instead.

TLDR: I hope that after this is fixed, it becomes impossible to be on a page without `?format=`, even if that page is only available in one format.

---

Another option instead of a redirect: if the user lands on a page available in more than one format (such as [Make your pages discoverable](https://amp.dev/documentation/guides-and-tutorials/optimize-and-measure/discovery/)), they are presented with a disambiguation page that forces them to select a format.

This ensures the user journey ensures the user is always aware of the format they are reading about. Otherwise, we still have the problem I mention in #3115 where e.g. someone interested in stories ends up on the "websites" filter and is presented with pages that aren't applicable to the format they really care about.

**Comment by tharders** [comment]
@fstanis You got it wrong. The idea is that a request without a format will never result in a format that is not supported. The first supported format is displayed. 
'websites' is only used as a default if supported.
For example a page that only supports email will be displayed in format email if no explicit format is specified.

Currently we always specify a canonical URL
So only the canonical url will end up in the search engine cache.

In most cases that is fine, since the content does not differ for the different supported formats.

Only a few pages uses the format filter which results in different content for different formats.
But I think in most cases the differences do not justify a separate version in the search index.
So for me it is fine to have only one format version in the search index.

**Comment by fstanis** [comment]
> differences do not justify a separate version in the search index

I fully agree with this. The issue I have is not with the content difference, but with UX.

Example user journey:

- User X is new to AMP.
- X wants to read about AMP for Email.
- X has never been on amp.dev before or at least not enough to be familiar with the color scheme or the fact there's a drop down in the top left corner.

1. X uses a search engine or finds a deep link on another website like StackOverflow. This link leads them to amp.dev.

2. The article they found is about an `amp-list`. Because this component is supported by websites, they are taken to the websites section.

3. X now sees all the components available on the left and clicks on `amp-audio`.

4. X proceeds to use `amp-audio` in an email, which doesn't work, as that component is not supported.

In this scenario, I see no reason X could have known they need to switch formats using the drop-down. This is explained *only* if they've been to [Get Started](https://amp.dev/documentation/) which explains which formats exist and the color scheme used.

If we want to have pages shared between different formats, we need to ensure deep links to them are meaningful for those new to AMP. Defaulting to websites is not sufficient in this regard.

**Comment by tharders** [comment]
Ok, makes sense.

I suggest we show a prominent info box with easy format toggle inside the page when more than one format is supported.
We would still have the default link as canonical page and the user will directly see the content of the page without the need to change the format if the default one is fine. 
The initial format should be selected by the logic explained above.

(We only have to think a little bit how we implement this info in regard to search engine optimization.)

@fstanis @sebastianbenz @CrystalOnScript @matthiasrohmer What do you think about the info box idea?

**Comment by sebastianbenz** [comment]
+1 I like the idea of having info box if more than one format is supported!

**Comment by fstanis** [comment]
+1 to info box idea.

Will it only be displayed if `?format=` is not specified? If so, I think it makes sense to keep the canonical links as-is (no `?format=`) and have the following behavior:

- If the user lands on a page with no `?format=` and the page supports more than one format: display info box along with first format in the list.
- If the user lands on a page with no `?format=` and the page supports **only one** format: redirect to `?format=<that format>`, e.g. `https://amp.dev/documentation/guides-and-tutorials/start/visual_story/` -> `https://amp.dev/documentation/guides-and-tutorials/start/visual_story/?format=stories`

Reason to prefer redirect over canonical: if we add a new format to a page (e.g. a new component now supported by email), the canonical doesn't change (it has no format), the page simply receives an infobox instead of a redirect.

**Comment by tharders** [comment]
@fstanis I think it is not a good idea to redirect a URL that is uses as canonical.
The important thing is that the page and the links in the page use the correct format.
There is no need to have the format parameter explicitly in the URL.
So we can simply show the page for the format that is supported without the info box if only one format is supported.
And all canonical URLs will still be without explicit parameter and there should be no problem when a new format is added for a page.

We will propose a layout for the info box in the next days.
